# Exercise 2
Add the specified code for each code cell, running the cells _in order_.

Open the included **`data/words_small.txt`** file (containing a "small" list of English words) and `read()` it into a single string variable **`words_str`**. Then print the first 100 _characters_ of the `words_str` variable (use a slice operator!)

In [1]:
with open('data/words_small.txt') as file:
    words_str = file.read()
print(words_str[0:100])

abbreviate
abbreviation
abdomen
able
aboard
abroad
absent
absolute
abuse
academic
academy
accelerate


**Using a loop**, count the number of times the newline character `\n` appears in the `words_str`. This should tell you the number of words in the file (plus or minus one). You can check your work with the built-in `count()` string method (but use the loop for practice!)

In [2]:
newline_count = 0
for letter in words_str:
    if letter == '\n':
        newline_count+=1
print(newline_count)
print(words_str.count('\n'))

1520
1520


Create a variable **`letter_counts`** that is a dictionary whose keys are characters (letters) and whose values are the number of times that character appears in `words_str`. Print the dictionary after you make it.
- To do this, start with an empty dictionary, then loop through the letters one at a time. At each letter, increase the value associated with that letter key by one.
- Watch out for if the letter is not in the dictionary yet!
- Also _delete_ the entry for the newline `\n` character

In [3]:
letter_counts = {}
for letter in words_str:
    # get the current number, or 0 if it doesn't exist
    current_count = letter_counts.get(letter, 0)
    
    # assign the new number
    letter_counts[letter] = current_count + 1

letter_counts.pop('\n')
print(letter_counts)

{'a': 703, 'b': 188, 'r': 642, 'e': 1118, 'v': 85, 'i': 515, 't': 604, 'o': 598, 'n': 550, 'd': 314, 'm': 223, 'l': 457, 's': 504, 'u': 279, 'c': 328, 'y': 189, 'p': 224, 'h': 295, 'f': 155, 'w': 161, 'g': 213, 'k': 126, 'z': 15, 'x': 23, 'j': 16, 'q': 9}


Print out all the count in the `letter_counts` dictionary in the format `"A appears 703 times"`, each on a new line, in alphabetical order. _Hint_: iterate through the sorted keys!

In [4]:
for letter in sorted(letter_counts):
    print(letter.upper(), "appears", letter_counts[letter], "times")

A appears 703 times
B appears 188 times
C appears 328 times
D appears 314 times
E appears 1118 times
F appears 155 times
G appears 213 times
H appears 295 times
I appears 515 times
J appears 16 times
K appears 126 times
L appears 457 times
M appears 223 times
N appears 550 times
O appears 598 times
P appears 224 times
Q appears 9 times
R appears 642 times
S appears 504 times
T appears 604 times
U appears 279 times
V appears 85 times
W appears 161 times
X appears 23 times
Y appears 189 times
Z appears 15 times


Print which letter from the `letter_counts` dictionary has a **value** of `515`. _Hint:_ use a loop to look through all the _items_ until you find the right one, then print that!

In [5]:
for letter,count in letter_counts.items():
    if count == 515:
        print(letter.upper(), "appears", letter_counts[letter], "times")
        break

I appears 515 times


Create a variable **`count_to_letter_dict`** that is a dictionary whose **keys** are the counts and whose **values** are the letters that have that count. This is a _reverse lookup_ dictionary. Hint: ceate a new dictionary, then loop through items in `letter_counts` and add an item to your new dictionary for each element. You can assume no duplicated counts.

In [6]:
count_to_letter_dict = {}
for letter,count in letter_counts.items():
    count_to_letter_dict[count] = letter  # assumes no duplicates, otherwise would need lists

Print the **10** most common letters and their counts with the format `"A appears 703 times"`. _Hint:_ you'll need to sort the keys to `count_to_letter_dict` and take the first 10 of those.

In [7]:
sorted_keys = sorted(count_to_letter_dict.keys(), reverse=True) 
for count in sorted_keys[:10]:  # slice!
    print(count_to_letter_dict[count].upper(), "appears", count, "times")

E appears 1118 times
A appears 703 times
R appears 642 times
T appears 604 times
O appears 598 times
N appears 550 times
I appears 515 times
S appears 504 times
L appears 457 times
C appears 328 times


Bonus: repeat your analysis on the `data/words_huge.txt` file! Are the top letters the same?